Introduce `summaryScreenReader` context to `cartItemPrice` filter.

peterwilsoncc opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

Introduces the filter `cartItemScreenReaderPrice` for modifying the screen-reader-only text in the mini cart on the checkout page. The filter includes the context `summary` to match other filters running on the checkout page.

The default text contains the placeholders `<quantity/>`, `<productName/>` and `<price/>`, these values must be included in the return value or an error will occur. Validation uses the function `productPriceScreenReaderValidation`.

Note: This was originally proposed as an additional context for the existing filter but after re-evaluation during the review process it was decided to use a new filter, see https://github.com/woocommerce/woocommerce/pull/62080#issuecomment-3756054765 below.

Closes #62009.
Closes WOOPLUG-5876
See https://linear.app/a8c/issue/WOODEP-61/a11y-due-today-text-incorrectly-hidden-from-screen-readers-on-checkout

#### Risks

~The big risk I can see with this change is that extensions making use of the filter will now change the screen reader text. Potentially in a way that they are not expecting or is unhelpful for users of assistive technology.~

~In the case of both Deposits and Subscriptions, the before and after text for the screen reader text is identical due to the way the filter checks the context (both use the same code, [see this code reference from Subs](https://github.com/woocommerce/woocommerce-subscriptions/blob/7ebb1eb8b2c0139fa4e39ea31a32cca7e4e015b7/client/core/filters/index.js#L107-L122)).~

Edit: These have been mitigated by the new filter name.

### How to test the changes in this Pull Request:

1. Enable multiple payment methods (this is to allow you to trigger React to re-render)
2. Configure the site to use block checkout
3. Add an item to the cart
4. Proceed to the checkout
5. Run the following code in your console `wc.blocksCheckout.registerCheckoutFilters( 'testing', { 'cartItemScreenReaderPrice': ( value, extensions, { context } ) => { value = 'replacement'; return value; } } );`
6. Change payment method to trigger re-render of page
7. Ensure the error is thrown due to the lack of replacement values in the filtered content
8. Reload the checkout page.
9. Run the following code in your console `wc.blocksCheckout.registerCheckoutFilters( 'testing', { 'cartItemScreenReaderPrice': ( value, extensions, { context } ) => { value = '<price/> <productName/> <quantity/> replacement'; return value; } } );`
10. Change the payment method to trigger re-render of page
11. Inspect the price element in the order summary
12. Ensure the associate screen reader text element shows the price (formatted with currency), product name and quantity followed by the word replacement
13. As a logic check, please run the coding sample provided in the documentation update to ensure it's bug free. As noted below I've done this but it would be good to check I haven't missed anything.


<img width="378" height="102" alt="Screenshot 2026-02-04 at 9 37 30 am" src="https://github.com/user-attachments/assets/ea8bdd3f-63c4-481f-8d04-95b10fe4f3cf" />


### Testing that has already taken place:

* Steps above
* code samples included in the documentation 

### Milestone

<!-- DO NOT remove or modify this section (other than to check the box). -->

<!-- milestone-target-selection -->
- [x] Automatically assign milestone for the **[next WooCommerce version](../blob/trunk/plugins/woocommerce/woocommerce.php#L6)**
<!-- /milestone-target-selection -->

> **Note:** Check the box above to have the milestone automatically assigned when merged.
> Alternatively (e.g. for point releases), manually assign the appropriate milestone.


### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [x] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

Introduce `cartItemScreenReaderPrice` checkout filter.

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

peterwilsoncc committed:
Introduce `summaryScreenReader` context to `cartItemPrice` filter.

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

peterwilsoncc commented:
@dkotter Are you able to take a look at this before I pass it on to the Woo team?

dkotter reviewed (commented):
This looks good to me. @vikrampm1 Do we want to pass this through our QA before assigning over to Woo?

vikrampm1 commented:
@ankitguptaindia @qasumitbagthariya if you can help with one round of QA here, we can pass it over to Woo for next steps.

peterwilsoncc committed:
Validate filter return value in screen reader context.

straku commented:
@peterwilsoncc @vikrampm1 My priority for this week, thanks for the ping 🙇

straku commented on plugins/woocommerce/client/blocks/packages/checkout/utils/validation/validate-filter.ts:
```diff
@@ -5,3 +5,19 @@ import mustContain from './must-contain';
+ * entirety.
+ */
+export const productPriceScreenReaderValidation = ( value: string ): true | never => {
+	if ( typeof value === 'string' && value.length > 0 ) {
```
I don't think we need to check whether we're dealing with string here. If anything other than a string were provided, an error would be thrown earlier, and we wouldn't reach the validation step at all ([see this logic](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/client/blocks/packages/checkout/filter-registry/index.ts#L231-L243)). TypeScript is right here in saying this is for sure a string.

Making sure a non-empty string is provided makes sense.

straku commented on plugins/woocommerce/client/blocks/packages/checkout/utils/validation/validate-filter.ts:
I can see the linter complaining about formatting here.

straku commented on plugins/woocommerce/client/blocks/packages/checkout/utils/validation/index.ts:
```diff
@@ -4,4 +4,4 @@ export {
 } from './get-validity-message-for-input';
 export { default as isPostcode } from './is-postcode';
-export { productPriceValidation } from './validate-filter';
+export { productPriceScreenReaderValidation, productPriceValidation } from './validate-filter';
```
This is also too long for linter.

straku commented:
@peterwilsoncc [This section of the documentation](https://github.com/woocommerce/woocommerce/blob/trunk/docs/block-development/extensible-blocks/cart-and-checkout-blocks/filters-in-cart-and-checkout/order-summary-items.md?plain=1#L108-L121) needs updating after this change.

Summary of needed changes courtesy of claude:

1. Add `summaryScreenReader` to the allowed values for `context`
2. Explain that `summaryScreenReader` is used for accessibility/screen reader text
3. Note that when context is `summaryScreenReader`:
    - The `defaultValue` is a full sentence (not `<price/>`)
    - The validation doesn't require `<price/>` - just a non-empty string
    - Extensions can provide completely custom screen reader text

peterwilsoncc committed:
CS Fixes.

peterwilsoncc committed:
Rename screen reader summary filter.

peterwilsoncc committed:
Convert summary to Interpolate Element.

peterwilsoncc committed:
Filter value must contain placeholders.

peterwilsoncc commented:
> I'm sorry that I originally asked to re-use an existing filter and now we've gone back on that

It's really no problem and there's no need to apologise. I've done exactly the same myself on more than one occasion.

I've pushed three commits:

* Introduced the filter `cartItemScreenReaderPrice`, it runs with the context `summary` to align with `cartItemPrice` even though it's not used in any other contexts at the moment
* Replaced the sprintf placeholders with `<productName/>` and `<quantity/>` and used `createInterpolateElement`
* Updated the validation function to require both of the placeholders be in the string

I _think_ I've got the coding standards correct this time, for some reason vscode is intermittent at best with the coding sniffs in this repo.

peterwilsoncc committed:
CS Fix: Line length.

straku commented on plugins/woocommerce/client/blocks/assets/js/base/components/cart-checkout/order-summary/order-summary-item.tsx:
```diff
@@ -2,6 +2,7 @@
  * External dependencies
  */
 import clsx from 'clsx';
+import { createInterpolateElement } from '@wordpress/element';
```
There's already an import from `@wordpress/element` in this file. `createInterpolateElement` should be moved there, linter is complaining.

straku commented on plugins/woocommerce/client/blocks/assets/js/base/components/cart-checkout/order-summary/order-summary-item.tsx:
```diff
@@ -119,6 +121,32 @@ const OrderSummaryItem = ( {
+		/* translators: %1$d is the total price including the currency symbol. <quantity/> and <productName/> are placeholders and should not be translated. */
+		_n(
+			'Total price for <quantity/> <productName/> item: %1$s',
+			'Total price for <quantity/> <productName/> items: %1$s',
```
Why are you not using `<price/>` for price instead of `%1$s`?

peterwilsoncc committed:
Move createInterpolateElement to existing import.

peterwilsoncc committed:
Replace price with placeholder.

peterwilsoncc committed:
Add price to required placeholders.

peterwilsoncc committed:
Update docblock following changes.

peterwilsoncc committed:
Document new filter `modifyCartItemScreenReaderPrice`.

woocommercebot committed:
Add changefile(s) from automation for the following project(s): woocommerce

peterwilsoncc committed:
Use _n to demonstrate plurals.

peterwilsoncc committed:
Typos, run on sentence.

woocommercebot committed:
Add changefile(s) from automation for the following project(s): woocommerce

straku committed:
Merge branch 'trunk' into fix/62009-filter-summary-screen-reader-text

straku commented on docs/block-development/extensible-blocks/cart-and-checkout-blocks/filters-in-cart-and-checkout/order-summary-items.md:
```diff
@@ -171,6 +172,95 @@ registerCheckoutFilters( 'example-extension', {
+
+> Filters can be also combined. See [Combined filters](/docs/block-development/extensible-blocks/cart-and-checkout-blocks/filters-in-cart-and-checkout/) for an example.
+
 ### Screenshots
```
This screenshot section applies to the `cartItemPrice` filter, not the `cartItemScreenReaderPrice`. Your additions need to sit below this section.

I know you've based your code examples on the structure that we have in `cartItemPrice`, but I don't agree that we need to repeat it. Since we're not conveying any new information with those examples, I think it would be better to keep it short and settle on the basic example (under one "Code example" section). What do you think?

peterwilsoncc committed:
Move screenshot back to original location.

peterwilsoncc committed:
Add missed optional chaining to docs.

straku committed:
Merge branch 'trunk' into fix/62009-filter-summary-screen-reader-text

straku committed:
Merge branch 'trunk' into fix/62009-filter-summary-screen-reader-text

straku committed:
Fix markdown tables

straku committed:
Merge branch 'trunk' into fix/62009-filter-summary-screen-reader-text

straku reviewed (approved):
Thanks for the contribution 👏
